lib/telemetry/telemetry.go

```diff
@@ -0,0 +1,90 @@
+package telemetry
+
+// PLSREVIEW(kit): Should this package be lib/telemetry, or should we define it at the top level alongside agent
+//  and lib?
+
+import (
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+type Label struct {
+	Key   string
+	Value string
+}
+
+// PLSREVIEW(kit): I decided to directly wrap the API of go-metrics because it informs all of our call sites.
+//  However, this does not need to be set in stone. For example, I renamed Incr to Inc. This is a good opportunity to
+//  adjust the metrics API we use internally. Any suggestions for improvements would be welcome!
+type MetricsClient interface {
+	// A Gauge should retain the last value it is set to
+	SetGauge(key []string, val float32, labels ...Label)
+
+	// KV should emit a Key/Value pair for each call.
+	EmitKV(key []string, val float32, labels ...Label)
+
+	// Counters accumulate
+	IncCounter(key []string, val float32, labels ...Label)
+
+	// Samples are for timing quantiles
+	AddSample(key []string, val float32, labels ...Label)
+	// Convenience fns to capture durations for samples
+	MeasureSince(key []string, start time.Time, labels ...Label)
+
+	// TODO(kit): I have not thought about filters yet. If we don't update them outside of RuntimeCfg
+	//  we don't need to offer an interface.
+
+	// TODO(kit): Should we have functions for rendering out the contents of the Inmemsink or arbitrary metrics state?
+	// 	That would allow us to use everything through a client, rather than having to keep a reference to and
+	//  interact with the sink in our agent. However, that could be more of a go-metrics implementation detail than
+	//  one that we can generalize. Maybe more research here is needed - or we just mirror go-metrics
+	//  and refactor the interface down the line if needed?
+	// Render()
+
+	// todo(kit): temporary value passthroughs. We want to delete these when we migrate Consul to import lib/telemetry
+	//  instead of go-metrics.
+	GetClient() interface{}
+	GetInmemSink() interface{}
```

Comment:
If we need to keep these around for a bit, I think you could have them return the concrete types (`*metrics.Metric`, and `*metrics.InmemSink`). It exposes a bit of the underlying implementation, but in the short term I think that will probably be easier and will avoid having to cast from `interface{}`.

Reply:
Agreed, this was an experiment with seeing if I could do away with any references to the internal client and memsink. I'm keeping the `GetInmemSink()` fn around because we do use the sink directly to render our non-prometheus aggregates. I adjusted it to return `*metrics.InmemSink` directly.